src/wxdoughnutchartctrl.cpp

```diff
@@ -24,28 +24,38 @@
 
 wxDoughnutChartCtrl::wxDoughnutChartCtrl(wxWindow *parent,
                                          wxWindowID id,
-                                         const wxDoughnutChartData &data,
+                                         wxDoughnutChartData &data,
                                          const wxPoint &pos,
-                                         const wxSize &size, 
+                                         const wxSize &size,
                                          long style)
     : wxChartCtrl(parent, id, pos, size, style),
     m_doughnutChart(data, size)
 {
+    data.AddObserver(this);
 }
 
-wxDoughnutChartCtrl::wxDoughnutChartCtrl(wxWindow *parent, 
-                                         wxWindowID id, 
-                                         const wxDoughnutChartData &data,
+wxDoughnutChartCtrl::wxDoughnutChartCtrl(wxWindow *parent,
+                                         wxWindowID id,
+                                         wxDoughnutChartData &data,
                                          const wxDougnutChartOptions &options,
                                          const wxPoint &pos,
-                                         const wxSize &size, 
+                                         const wxSize &size,
                                          long style)
     : wxChartCtrl(parent, id, pos, size, style),
     m_doughnutChart(data, options, size)
 {
+    data.AddObserver(this);
 }
 
 wxDoughnutChart& wxDoughnutChartCtrl::GetChart()
 {
     return m_doughnutChart;
 }
+
+void wxDoughnutChartCtrl::OnUpdate(const std::unordered_map<wxString,wxChartSliceData> &data)
+{
+    m_doughnutChart.SetData(data);
+    auto parent = this->GetParent();
```

Comment:
Why are we calling Layout() here? The size of the control shouldn't change because we are adding data.
We should however call Refresh() to indicate that the control needs to be repainted.

Reply:
For some reason it isn't work for me even with [Update/0](http://docs.wxwidgets.org/trunk/classwx_window.html#abaf28f1a075fd1b10f761a8febe597ec)

Reply:
Did you call Refresh() before Update() ?

Reply:
Yes, as it's mentioned in the documentation.